README.md
# ng-fire-alarm
> Distributed via

[![Version     ](https://img.shields.io/gem/v/ng-fire-alarm.svg)                               ](https://rubygems.org/gems/ng-fire-alarm)
[![Bower Version](https://badge.fury.io/bo/ng-fire-alarm.png)                               ](https://badge.fury.io/bo/ng-fire-alarm)

> Firebase binding use $q in AngularJS

[![Travis CI   ](https://img.shields.io/travis/tomchentw/ng-fire-alarm/master.svg)             ](https://travis-ci.org/tomchentw/ng-fire-alarm)
[![Quality     ](https://img.shields.io/codeclimate/github/tomchentw/ng-fire-alarm.svg)        ](https://codeclimate.com/github/tomchentw/ng-fire-alarm)
[![Coverage    ](https://img.shields.io/coveralls/tomchentw/ng-fire-alarm.svg)                 ](https://coveralls.io/r/tomchentw/ng-fire-alarm)
[![Dependencies](https://img.shields.io/gemnasium/tomchentw/ng-fire-alarm.svg)                 ](https://gemnasium.com/tomchentw/ng-fire-alarm)


## Project philosophy

### Develop in LiveScript
[LiveScript](http://livescript.net/) is a compile-to-js language, which provides us more robust way to write JavaScript.  
It also has great readibility and lots of syntax sugar just like you're writting python/ruby.

### Use new API from $q
We use newly introduced $q API to [`notify`](https://github.com/angular/angular.js/blob/master/CHANGELOG.md#120rc1-spooky-giraffe-2013-08-13) you about the *value/order* changes, we let you decide how you deal with your data with `$scope`.

### Firebase Collection support
We know that you want to take advantage of collection in `Firebase`, but still want to preserve the right order, or order by any properties in each item. `ng-fire-alarm` allows to transform collection **object** into native js array for you.

### Seperation of Concerns
We follow seperation of concerns by seperating control(reference) object and data source to [`Alarm`](https://github.com/tomchentw/ng-fire-alarm#alarm-object) object and [`Fire`](https://github.com/tomchentw/ng-fire-alarm#fire-objects) object(s). So a `ng-change`, `ng-submit`, `ng-click` can be directly bound to Alarm object, while your Fire object(s) are clean and just like what you see in Firebase dashboard.


## Installation

### Just use it

* Download and include [`ng-fire-alarm.js`](https://github.com/tomchentw/ng-fire-alarm/blob/master/ng-fire-alarm.js) OR [`ng-fire-alarm.min.js`](https://github.com/tomchentw/ng-fire-alarm/blob/master/ng-fire-alarm.min.js).  

Then include them through script tag in your HTML.

### **Rails** projects (Only support 3.1+)
Add this line to your application's Gemfile:
```ruby
gem 'ng-fire-alarm'
```

And then execute:

    $ bundle

Then add these lines to the top of your `app/assets/javascripts/application.js` file:

```javascript
//= require angular
//= require ng-fire-alarm
```

And include in your `angular` module definition:
     
    var module = angular.module('my-awesome-project', ['ng-fire-alarm']).


## Usage

We add a new method to `Firebase.prototype`:

### $toAlarm

The key to transform your `Firebase` reference into a `AngularJS` powered [alarm object](https://github.com/tomchentw/ng-fire-alarm/blob/master/README.md#alarm-object), it take one parameter:

#### Options

##### collection: _true/otherwise_.

Pass true will transform collection object into native js array for you.

#### `Alarm` object
A wrapped object over Firbease reference that is returned by calling `Firebase.prototype.$toAlarm`. 

##### Attributes:

* `$promise`: it will be notified everytime the value/order changes.

##### Convenience Method:

* `$thenNotify`: Same as `$promise.then(void, void, callback)`.  
Register a callback that notify you each time the alarm rings.  
Unlike `$promise.then`, this method is self-chained and will return alarm object to allow chaining.  
See examples below.

##### Query Methods:

Ther're wrapper for `Firebase.prototype.limit/startAt/endAt` function, but it'll update internal Firebase reference and it'll populate new data through your callback registered via `$thenNotify`.

* [$limit](https://www.firebase.com/docs/javascript/firebase/limit.html)
* [$startAt](https://www.firebase.com/docs/javascript/firebase/startat.html)
* [$endAt](https://www.firebase.com/docs/javascript/firebase/endat.html)

Examples:

```JavaScript
var ROOT = new Firebase('https://ng-fire-alarm.firebaseio.com/');

function UsersListCtrl ($scope) {
  $scope.users = [];

  $scope.usersAlarm = ROOT.child('users').$toAlarm({collection: true});

  $scope.usersAlarm.$thenNotify(function(users) {
    $scope.users = users;
  });
}
```

Fast prototyping!

```HTML
<ul ng-controller="UsersListCtrl" infinite-scroll="usersAlarm.$limit(users.length + 25)">
  <li ng-repeat="user in users">
    {{ user.name }}
  </li>
</ul>
```

##### Write Methods:
They're wrapper for `Firebase.prototype.remove/push/update/set/setPriority/setWithPriority` function, but it'll return a `promise` object instead of passing in a callback function.

* [$remove](https://www.firebase.com/docs/javascript/firebase/remove.html)
* [$push](https://www.firebase.com/docs/javascript/firebase/push.html)
* [$update](https://www.firebase.com/docs/javascript/firebase/update.html)
* [$set](https://www.firebase.com/docs/javascript/firebase/set.html)
* [$setPriority](https://www.firebase.com/docs/javascript/firebase/setpriority.html)
* [$setWithPriority](https://www.firebase.com/docs/javascript/firebase/setwithpriority.html)

Examples:

```JavaScript
var ROOT = new Firebase('https://ng-fire-alarm.firebaseio.com/');

function UserEditCtrl ($scope) {
  $scope.userAlarm = ROOT
    .child('users/`')
    .$toAlarm()
    .$thenNotify(function(user) {  $scope.user = user;  });
}
```

Fast prototyping!

```HTML
<form ng-controller="UserEditCtrl">
  <input type="text" ng-model="user.name" ng-change="userAlarm.$update(user)">
</form>
```


#### `Fire` object(s)
Object that is passed in to callbacks registered via `$thenNotify`, they can be **primitive, object, or array**:

##### primitive

Primitive is just js primitive. There's **NO** **NO** **NO** wrapper around primitive <del>`{$value: primitive}`</del>.

```JavaScript
bell.$thenNotify(function (aStringOrANumber) { $scope.myVar = aStringOrANumber; });
```

##### object
  
we've add two properties on it:  

  1. [`$name`](https://www.firebase.com/docs/javascript/datasnapshot/name.html)
  2. [`$priority`](https://www.firebase.com/docs/javascript/datasnapshot/getpriority.html)

##### array

sorted by native Firebase [ordering](https://www.firebase.com/docs/javascript/firebase/setpriority.html).  

  Each item in array, if they're object, will have these properties:

  1. [`$name`](https://www.firebase.com/docs/javascript/datasnapshot/name.html)
  2. [`$priority`](https://www.firebase.com/docs/javascript/datasnapshot/getpriority.html)
  3. [`$index`](https://www.firebase.com/docs/javascript/datasnapshot/foreach.html): object index in array. Useful for reverse ordering  

```HTML
<div ng-repeat="item in array | orderBy:'$index':true"></div>
```


## Contributing

[![devDependency Status](https://david-dm.org/tomchentw/ng-fire-alarm/dev-status.svg?theme=shields.io)](https://david-dm.org/tomchentw/ng-fire-alarm#info=devDependencies)

1. Fork it ( https://github.com/tomchentw/ng-fire-alarm/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
<a name="0.4.6"></a>
### 0.4.6 (2014-03-14)


#### Bug Fixes

* typo and exclude Makefile in bower ([1ffa62d4](https://github.com/tomchentw/ng-fire-alarm/commit/1ffa62d435e02f2d45621f37afa3d920a45fa4ab))
* **README.md:** update coveralls link ([be57bef9](https://github.com/tomchentw/ng-fire-alarm/commit/be57bef90025f1b5baffba33c9b01c6599ee806f))


<a name="0.4.5"></a>
### 0.4.5 (2014-03-09)


<a name="0.4.4"></a>
### 0.4.4 (2014-02-24)


<a name="0.4.3"></a>
### 0.4.3 (2014-02-23)


<a name="0.4.2"></a>
### 0.4.2 (2014-02-09)


#### Bug Fixes

* **AlarmReceiver:** setTimeout before startWatching ([f362c0cc182f0aac4813ac19f5b43f1f8fdd678a](https://github.com/tomchentw/ng-fire-alarm/commit/f362c0cc182f0aac4813ac19f5b43f1f8fdd678a))
* **FireAlarm:** hide _alarmReceiver with closure function ([288700f6a11440517c70629d8ee04ba76768bde8](https://github.com/tomchentw/ng-fire-alarm/commit/288700f6a11440517c70629d8ee04ba76768bde8))


<a name="0.4.1"></a>
### 0.4.1 (2014-02-08)


#### Features

* **gulpfile:** adjust for gulp 3.5.0 from tomchentw-boilerplate ([6a7b47dc83ecab12282b150f20ac0e127ec184d8](https://github.com/tomchentw/ng-fire-alarm/commit/6a7b47dc83ecab12282b150f20ac0e127ec184d8))
* **ng-fire-alarm:** cleaner APIs ([3a13c354fea1ad64a171c14a4b9850247b0e8765](https://github.com/tomchentw/ng-fire-alarm/commit/3a13c354fea1ad64a171c14a4b9850247b0e8765))


<a name="0.3.4"></a>
### 0.3.4 (2014-01-18)


#### Bug Fixes

* **$fireAlarm:** second argument should be Array or something else ([57adeb03e58820855c1ed2c13fb1326a37fbd116](https://github.com/tomchentw/ng-fire-alarm/commit/57adeb03e58820855c1ed2c13fb1326a37fbd116))


<a name="0.3.3"></a>
### 0.3.3 (2014-01-17)


#### Bug Fixes

* **gulpfile:** add release-git task with `git push`, `--tags` ([cdee4e314d2e3404be3ec3ac53957cc050894934](https://github.com/tomchentw/ng-fire-alarm/commit/cdee4e314d2e3404be3ec3ac53957cc050894934))


<a name="0.3.2"></a>
### 0.3.2 (2014-01-17)


#### Bug Fixes

* **rubygems:** rename to correct module name ([039ced1b74b573d8d6b79fe4619d269030b9ef57](https://github.com/tomchentw/ng-fire-alarm/commit/039ced1b74b573d8d6b79fe4619d269030b9ef57))


<a name="0.3.1"></a>
### 0.3.1 (2014-01-17)


#### Features

* **ng-fire-alarm:** rename repo and introduce new APIs ([71499ea03771ee0905d2cee936fb08a0f684e5d8](https://github.com/tomchentw/ng-fire-alarm/commit/71499ea03771ee0905d2cee936fb08a0f684e5d8))



ng-fire-alarm.js
/*! ng-fire-alarm - v 0.4.6 - Fri Mar 14 2014 12:03:06 GMT+0800 (CST)
 * https://github.com/tomchentw/ng-fire-alarm
 * Copyright (c) 2014 [tomchentw](https://github.com/tomchentw);
 * Licensed [MIT](http://tomchentw.mit-license.org)
 */
/*global angular:false, Firebase:false*/
(function(){
  var isObject, AlarmReceiver, Fireman, Firemen, FireAlarm, slice$ = [].slice;
  isObject = angular.isObject;
  function assignNamePriority(dataSnap, it){
    if (isObject(it)) {
      it.$name = dataSnap.name();
      it.$priority = dataSnap.getPriority();
    }
    return it;
  }
  function buildNgObject(dataSnap, singlecton){
    var val;
    val = dataSnap.val();
    if (null === val) {
      return assignNamePriority(dataSnap, singlecton);
    }
    if (!isObject(val)) {
      return val;
    }
    assignNamePriority(dataSnap, val);
    if (isObject(singlecton)) {
      return angular.extend(singlecton, val);
    } else {
      return val;
    }
  }
  function buildDeferFunctor(defer){
    return function(it){
      if (it) {
        defer.reject(it);
      } else {
        defer.resolve();
      }
    };
  }
  AlarmReceiver = (function(){
    AlarmReceiver.displayName = 'AlarmReceiver';
    var prototype = AlarmReceiver.prototype, constructor = AlarmReceiver;
    AlarmReceiver.create = function(query, defer, options){
      var ctor;
      ctor = true === options.collection ? Firemen : Fireman;
      return new ctor(query, defer, options);
    };
    function AlarmReceiver(_query, _defer, options){
      var this$ = this;
      this._query = _query;
      this._defer = _defer;
      this._isSingleton = true === options.singlecton;
      this._singlecton = void 8;
      setTimeout(function(){
        this$.startWatching();
      });
    }
    prototype.update = function(method, it){
      var _query;
      _query = this._query;
      _query.on('value', angular.noop, angular.noop, constructor);
      this.stopWatching();
      this._query = _query[method](it);
      this.startWatching();
      return _query.off('value', void 8, constructor);
    };
    prototype.notify = function(it){
      this._defer.notify(this._isSingleton ? this._singlecton : it);
    };
    prototype.onError = function(it){
      this._defer.reject(it);
    };
    return AlarmReceiver;
  }());
  Fireman = (function(superclass){
    var prototype = extend$((import$(Fireman, superclass).displayName = 'Fireman', Fireman), superclass).prototype, constructor = Fireman;
    prototype.startWatching = function(){
      this._query.on('value', this.onValue, this.onError, this);
    };
    prototype.stopWatching = function(){
      this._query.off('value', void 8, this);
    };
    prototype.onValue = function(dataSnap){
      var ngObject;
      ngObject = buildNgObject(dataSnap, this._singlecton);
      if (this._isSingleton && !this._singlecton) {
        this._singlecton = ngObject;
      }
      this.notify(ngObject);
    };
    function Fireman(){
      Fireman.superclass.apply(this, arguments);
    }
    return Fireman;
  }(AlarmReceiver));
  Firemen = (function(superclass){
    var prototype = extend$((import$(Firemen, superclass).displayName = 'Firemen', Firemen), superclass).prototype, constructor = Firemen;
    function Firemen(){
      Firemen.superclass.apply(this, arguments);
      this._isSingleton = true;
      this._singlecton = [];
      this._names = {};
    }
    prototype.notify = function(){
      var that;
      if (that = this._singlecton[0]) {
        this._singlecton[0] = JSON.parse(
        JSON.stringify(
        that));
      }
      superclass.prototype.notify.call(this);
    };
    prototype.startWatching = function(){
      this._query.on('child_added', this.onChildChanged, this.onError, this);
      this._query.on('child_changed', this.onChildChanged, this.onError, this);
      this._query.on('child_moved', this.onChildChanged, this.onError, this);
      this._query.on('child_removed', this.onChildRemoved, this.onError, this);
    };
    prototype.stopWatching = function(){
      this._query.off('child_added', void 8, this);
      this._query.off('child_changed', void 8, this);
      this._query.off('child_moved', void 8, this);
      this._query.off('child_removed', void 8, this);
    };
    prototype.rebuildNameIndex = function(start, end){
      var _singlecton, _names, i$, to$, i, item;
      _singlecton = this._singlecton, _names = this._names;
      for (i$ = start, to$ = end || _singlecton.length; i$ < to$; ++i$) {
        i = i$;
        item = _singlecton[i];
        if (isObject(item)) {
          item.$index = _names[item.$name] = i;
        }
      }
    };
    prototype.indexOf = function(name, del){
      var _names, index;
      _names = this._names;
      if (!(name in _names)) {
        return -1;
      } else {
        index = _names[name];
        if (del) {
          delete _names[name];
        }
        return index;
      }
    };
    prototype.onChildChanged = function(childSnap, prevName){
      var _singlecton, name, curIndex, childSinglection, ngIndex;
      _singlecton = this._singlecton;
      name = childSnap.name();
      curIndex = this.indexOf(name, true);
      if (curIndex !== -1) {
        childSinglection = _singlecton.splice(curIndex, 1)[0];
      }
      ngIndex = this._names[name] = 1 + this.indexOf(prevName);
      _singlecton.splice(ngIndex, 0, buildNgObject(childSnap, childSinglection));
      this.rebuildNameIndex(ngIndex);
      this.notify();
    };
    prototype.onChildRemoved = function(oldChildSnap){
      var curIndex;
      curIndex = this.indexOf(oldChildSnap.name(), true);
      this._singlecton.splice(curIndex, 1);
      this.rebuildNameIndex(curIndex);
      this.notify();
    };
    return Firemen;
  }(AlarmReceiver));
  FireAlarm = (function(){
    FireAlarm.displayName = 'FireAlarm';
    var QUERY_METHODS, WRITE_METHODS, prototype = FireAlarm.prototype, constructor = FireAlarm;
    FireAlarm.$q = void 8;
    function FireAlarm($promise, _ar){
      this.$promise = $promise;
      this._ar = function(){
        return _ar;
      };
    }
    prototype.$query = function(){
      return this._ar()._query;
    };
    prototype.$ref = function(){
      return this.$query().ref();
    };
    QUERY_METHODS = ['limit', 'startAt', 'endAt'];
    angular.forEach(QUERY_METHODS, function(name){
      prototype["$" + name] = function(it){
        this._ar().update(name, it);
      };
    });
    WRITE_METHODS = ['remove', 'push', 'update', 'set', 'setPriority', 'setWithPriority'];
    angular.forEach(WRITE_METHODS, function(name, index){
      var sliceAt;
      sliceAt = (function(){
        switch (index) {
        case 0:
          return 0;
        case 5:
          return 2;
        default:
          return 1;
        }
      }());
      prototype["$" + name] = function(){
        var args, defer, ref$;
        args = slice$.call(arguments, 0, sliceAt);
        defer = constructor.$q.defer();
        args.push(buildDeferFunctor(defer));
        (ref$ = this.$ref())[name].apply(ref$, args);
        return defer.promise;
      };
    });
    prototype.$thenNotify = function(it){
      this.$promise = this.$promise.then(void 8, void 8, it);
      return this;
    };
    return FireAlarm;
  }());
  angular.module('ng-fire-alarm', []).value('Firebase', Firebase).run(['$q', 'Firebase'].concat(function($q, Firebase){
    FireAlarm.$q = $q;
    return Firebase.prototype.$toAlarm = function(options){
      var defer, alarmReceiver;
      options || (options = {});
      defer = $q.defer();
      alarmReceiver = AlarmReceiver.create(this, defer, options);
      return new FireAlarm(defer.promise, alarmReceiver, options);
    };
  }));
  function extend$(sub, sup){
    function fun(){} fun.prototype = (sub.superclass = sup).prototype;
    (sub.prototype = new fun).constructor = sub;
    if (typeof sup.extended == 'function') sup.extended(sub);
    return sub;
  }
  function import$(obj, src){
    var own = {}.hasOwnProperty;
    for (var key in src) if (own.call(src, key)) obj[key] = src[key];
    return obj;
  }
}).call(this);
